### PR TITLE
container: add container placement verification

### LINKED ISSUE #413: Verification function for container-level thing
## Is your feature request related to a problem? Please describe.
I'm always frustrated when we can't check something container-level.

## Describe the solution you'd like
Smth. like `VerifySignature(cid interop.Hash256, msg []byte, sigs []Signature, curveHash NamedCurveHash) bool`, we need to:
 * use https://pkg.go.dev/github.com/nspcc-dev/neo-go/pkg/interop/native/crypto#VerifyWithECDsa internally
 * implement M/N internally precisely following the container placement policy (#412), each REP X should be satisfied by one of the appropriate nodes

This can be used both in contracts and verification scripts (with CID fixed). A set of signatures like this can prove object existence (refs. https://github.com/nspcc-dev/neofs-api/issues/299).

### PR DESCRIPTION
**Pull request by carpawell**
The new `VerifyPlacementSignatures` checks if a message was signed by a suitable number of nodes. Number is taken from the contract storage, their relevance should be ensured and maintained by the Alphabet nodes every epoch. Closes #413.

### CODE REVIEW

**Comment by carpawell** [comment]
Not the most optimal implementation but good enough for now.

Not sure if it should have `curveHash` arg. I think no.

File: contracts/container/contract.go [hunk 1]
```diff
@@ -505,6 +505,8 @@ const maxNumOfREPs = 255
 // only when a container list is changed, otherwise nothing should be done.
 // Call must be signed by the Alphabet nodes.
 func AddNextEpochNodes(cID interop.Hash256, placementVector uint8, publicKeys []interop.PublicKey) {
+	runtime.Log("fuck")
```
Comment (roman-khimov) [inline contracts/container/contract.go:508]: Nice one.

**Review by roman-khimov — COMMENTED** [review]
Looks legit for unordered signatures.

File: contracts/container/contract.go (hunk 1 above)
Comment (carpawell) [reply contracts/container/contract.go]: \*UFO landed and took this line\*

File: contracts/container/contract.go [hunk 2]
```diff
@@ -582,6 +582,48 @@ func validatePlacementIndex(ctx storage.Context, cID interop.Hash256, inx uint8)
 	}
 }
 
+// VerifyPlacementSignatures verifies that message has been signed by container
+// members according to container's placement policy: there should be at least
+// REP number of signatures for every placement vector. sigs must be container's
+// number of SELECTs length.
+func VerifyPlacementSignatures(cid interop.Hash256, msg []byte, sigs [][]interop.Signature, curveHash crypto.NamedCurveHash) bool {
```
Comment (roman-khimov) [inline contracts/container/contract.go:589]: Let's drop `curveHash` and simplify interface here. We're 256r1 only for this.

Comment (roman-khimov) [reply contracts/container/contract.go:589]: I also had some thought on special `msg` for verification scripts, but maybe we can leave it as is. `msg` can be created like here in the verification script: https://github.com/nspcc-dev/neo-go/blob/a9242535db757faba860e21c194033f30c679f48/pkg/core/native/native_test/cryptolib_verification_test.go#L173 and then only `sigs` are a part of invocation script (as they should be).

File: contracts/container/contract.go [hunk 3]
```diff
@@ -582,6 +582,48 @@ func validatePlacementIndex(ctx storage.Context, cID interop.Hash256, inx uint8)
...
+			if counter == m {
+				i++
+				continue repsLoop
+			}
+		}
+
+		panic("verified " + std.Itoa10(counter) + " signatures for " + std.Itoa10(i) + " placement vector, required: " + std.Itoa10(m))
+	}
+
+	return true
```
[the remark this answers is not in the document]
Comment (roman-khimov) [reply contracts/container/contract.go:624]: For various reuse cases (like calling from the other contract) `false` is better. It's also more natural for verification scripts. Panics are easier to debug though, you just have more data this way. Let's make it `return false`.

[the remark this answers is not in the document]
Comment (carpawell) [reply contracts/container/contract.go:624]: @roman-khimov, so what is the best approach? always `true` or `false` and never panic? or only the last `panic` make `false`? without panic test cases are a little bit worse then

[the remark this answers is not in the document]
Comment (roman-khimov) [reply contracts/container/contract.go:624]: In this particular case --- never panic. Most of `safe` methods don't panic.

[the remark this answers is not in the document]
Comment (carpawell) [reply contracts/container/contract.go:624]: ok, "most" is always a good point, however, i would panic in such cases

**Review by roman-khimov — APPROVED** [review]

**Pull request merged**